### PR TITLE
spec: add flow

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

Add flow charts. 

### Open Questions

- when a users account is in the red they will need to follow a similar flow to liquidation, but it would sell of a smaller portion of the obligation. Should the user mark close the trade then notify a liquidation bot to handle closing the position? This is something I don't fully understand. 

# [rendered](https://github.com/marbar3778/Entropy/blob/marko/flows/spec/flow.md)

### CODE REVIEW

File: spec/margin.md [hunk 1]
```diff
@@ -86,6 +84,18 @@ Deposits funds into an obligation account to be used to open a leveraged trade.
...
+    ///   1. `[]` Borrow reserve account.
+    ///   2. `[writable]` Obligation
+    ///   3. `[writable]` Obligation token mint
+    ///   4. `[writable]` Obligation token output
+    ///   5. `[]` Obligation token owner
+    ///   6. `[]` Lending market account.
+    ///   7. `[]` Derived lending market authority.
+
+    let deposit_reserve_info = next_account_info(account_info_iter)?;
+    let borrow_reserve_info = next_account_info(account_info_iter)?;
```
Comment (austinabell) [inline spec/margin.md:97]:
```suggestion
```

Comment (marbar3778) [reply spec/margin.md:97]: oh whoops, I was working on something else and this got pushed. will revert.

File: Cargo.toml [hunk 2]
```diff
@@ -1,4 +0,0 @@
-[workspace]
-members = [
-    "programs/*"
-]
```
[the remark this answers is not in the document]
Comment (marbar3778) [reply Cargo.toml]: my env was doing something and it resulted in this, reverted now

**Review by austinabell — APPROVED** [review]

**Comment by cmwaters** [comment]
> when a users account is in the red they will need to follow a similar flow to liquidation, but it would sell off a smaller portion of the obligation. Should the user mark close the trade then notify a liquidation bot to handle closing the position? This is something I don't fully understand.

Yeah I guess so, or the user could do it as well. We give rewards for someone successfully executing a liquidation so it doesn't matter if it is the user as part of the close position or a liquidation bot. 

I was thinking it might be better if, when the user closes a position in the red, that we always use all of the obligation to pay back the loan and thus only a portion of the tokens in the settle account. (i.e. the opposite of what you are saying)

(I find this such an annoying UX that you're constantly relying on these bots to try fulfil multiple parts of the contract and thus having to pay them fees in order to incentivise them)

File: spec/flow.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+## Open
+
+Open starts with a user setting a leverage level, borrows the funds needed, then creates a new order. At this stage the account is marked as having an open trade, prohibiting an withdrawals until the loan has been repaid in full with interest. 
+
+![Open](./assets/open.png)
+
+## Close
+
+Close begins with first creating a new order on the serum dex to close the open trade.
```
Comment (cmwaters) [inline spec/flow.md:23]: For close, do we allow a limited order or just have `immediate or cancel` order type

File: spec/flow.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+![Close](./assets/close.png)
+
+## Withdraw
+
+Withdraw settles the funds by calling the settle function on the serum dex. Once it has received the funds it will repay the loan. In the process of repaying the loan, if in the profit, the obligation will be sent to the user.
```
Comment (cmwaters) [inline spec/flow.md:29]: When you say sends funds to program account do you mean the margin account

Comment (cmwaters) [reply spec/flow.md:29]: Also isn't it the `settle` call first made through the margin account which then calls `settle` on the serum DEX

File: spec/flow.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+![Withdraw](./assets/withdraw.png)
+
+## Liquidate
```
Comment (cmwaters) [inline spec/flow.md:33]: Aren't some of these calls, i.e. the calls involved in taking the money from the settle account, trading it back for the original denomination, moving it to the margin account, then finally repaying a portion of the loan, the same calls as with the close position

File: spec/flow.md (hunk 4 above)
Comment (cmwaters) [reply spec/flow.md:29]: We may need to also sell some of the funds from the obligation account

**Review by cmwaters — COMMENTED** [review]
Nice, I think this makes it a lot clearer. I think it might be helpful now to go one step further and outline the validation rules for each of these transactions i.e. check that the user has enough collateral, check the price of the trading pair on the dex and so forth

File: spec/flow.md (hunk 5 above)
Comment (marbar3778) [reply spec/flow.md:33]: that would be correct, will make amendments prior to merge

**Comment by marbar3778** [comment]
Going to merge this and amend later

**Pull request merged**